### ISSUE TITLE
Standardize the `oid/future/data/...` property of the `<StorageFormat>On<Engine>DataframePartition` class

### ISSUE THREAD

**Issue by prutskov** [modin-project/modin#4530]
**Is your feature request related to a problem? Please describe.**

Currently, we have a different APIs for property of physical data stored by specific partition class.
https://github.com/modin-project/modin/blob/dcee13d57ebf9a006460deedb734c15791acae7a/modin/core/execution/dask/implementations/pandas_on_dask/partitioning/partition.py#L45
https://github.com/modin-project/modin/blob/dcee13d57ebf9a006460deedb734c15791acae7a/modin/core/execution/python/implementations/pandas_on_python/partitioning/partition.py#L45
https://github.com/modin-project/modin/blob/dcee13d57ebf9a006460deedb734c15791acae7a/modin/core/execution/ray/implementations/cudf_on_ray/partitioning/partition.py#L55
https://github.com/modin-project/modin/blob/dcee13d57ebf9a006460deedb734c15791acae7a/modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition.py#L54
https://github.com/modin-project/modin/blob/dcee13d57ebf9a006460deedb734c15791acae7a/modin/experimental/core/execution/native/implementations/omnisci_on_native/partitioning/partition.py#L61-L63

We have a number of possible problems:
 1. Different name of property doesn't allow to create a unified implementation over all types of partitions. For example, case when we should provide other partition to args of `apply` call. Current logic of partition class doesn't expect partition type in args/kwargs. Therefore, we should provide physical data to args:
```python
class PandasDataframePartitionManager:
   ...
  def some_unified_functionality(cls, parts_left, parts_right):
      ...
      [lpart.apply(rpart.<oid/future/data/...>)
         for lpart, rpart in zip(parts_left, parts_right)]  # --> We should implement this logic for every engine, that introduced a lot of code duplication
```
 2.  Physical data can be corrupted because the property is public and can be changed outside after `__init__` during life of partition object (property setter issue).
 3. Physical data should be provided to outside only after `self.drain_call_queue` call. Because without executing of all operations in call queue we will get incorrect physical data (property getter issue).

The possible solutions:

1. API solutions:
- a) Standardize physical data property. Need to have the single API for acces to physical data for any partition type.
OR
- b) Add logic of expecting of partition type in `apply/add_to_apply_calls`. The soultion, possible, isn't good because in that case we should check every argument of `apply, add_to_apply_calls` on `isinstance(arg, partition_type)`, it will introduce performance gaps.
2. Setter issue solution:
- Physical data property should be readonly to exclude unexpected changing outside. 
3. Getter issue solution:
- Physical data property should include `drain_call_queue` logic before providing physical data to outside to be up to date

The concept from ideas above:
```python
class <StorageFormat>On<Engine>DataFramePartition:
   def __init__(self, physical_data, ...):
      self._physical_data = physical_data
   
   @property
   def physical_data(self):
      self.drain_call_queue()
      return self._physical_data 
```

@modin-project/modin-core, @modin-project/modin-contributors,  what're you thoughts?

**Comment by anmyachev** [comment]
+1 that this need to be resolved.

Maybe https://github.com/modin-project/modin/pull/4196/files will be useful to you. It offers a solution through a property.

**Comment by prutskov** [comment]
Reopened because only problem 1 from the issue description was resolved by #4563

**Comment by mvashishtha** [comment]
Besides the call queue requiring draining, another problem with accessing `_data` is that non-full-axis virtual partitions don't have any `_data` property, but they are supposed to expose the same API as block partitions. That's the cause of binary operation failure #4691.  Likewise, methods like [`PandasOnDaskDataframePartitionManager. wait_partitions`](https://github.com/modin-project/modin/blob/7a36071c0b00e0392615a0dd9d5c2ddd5f7c0d27/modin/core/execution/dask/implementations/pandas_on_dask/partitioning/partition_manager.py#L55) will error with virtual partitions.
